### PR TITLE
submit higher priority jobs regardless the threshold

### LINKED ISSUE #8129: Job submitter bottleneck
Let me put this in a GH issue as I think this is going (was) to be impacting protection.
With large number of low priority (63k prio) pending on all agents, higher priority work WQE does get acquired, the jobs are created (regardless of thresholds according to @amaltaro ) but then jobs do not get submitted over threshold.
Example an agent with 40k job pending threshold could be holding 35k low priority job and de-facto is reduced to a 5k effective threshold for all the rest. That is, the low priority pending have no chance at running (because of higher priority work being ran otherwise from other agents) and will not free their pending slot, leaving very little room for other work that would deserve to get submitted and ran.

we should find a way to solve this

### PR DESCRIPTION
**Pull request by ticoann**
fixes #8129, #8284
The simplest way I can think of to solve this issue. (There are short comings though)

1. It will still submit higher priority workflow but only the highest one first.
   i.e. current highest priority is 5 and there are jobs in 7,8,9 in the wmbs queue. It will submit priority 9 only then once 9 is done, 8, etc
2. Not all the jobs with 9 will be submitted. only 20% of current slot. (This can be increased) - maybe will make configurable.

### CODE REVIEW

**Comment by ticoann** [comment]
@amaltaro, Alan please review. still need to combine Production and Processing priority but might do it with separate patch.

**Comment by amaltaro** [comment]
Seangchan, did you add this new key `wf_highest_priority` just for the cases where the JobSubmitter cache is stale?

JobSubmitter already makes sure to submit the highest priority jobs first:
https://github.com/dmwm/WMCore/blob/master/src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py#L539

and what we discussed was to simply remove the Production/Processing thresholds, not to over-submit based on the workflow priority (and respecting a factor that will never make everyone happy :)).

I think we can keep it simpler by implementing what is in https://github.com/dmwm/WMCore/issues/8267 , so, if the submitter passes the restrictions, we allow job submission. Of course, we still want to control the IO jobs.

**Comment by ticoann** [comment]
Hi Alan, I must misunderstood. I thought removing Production/Processing is in addition to this change. The problem I see is that we can always have the same problem not submitting the jobs in JobSubmitter if we don't by pass the threshold limit for the higher priority. We can chat if you have a time.

**Comment by ticoann** [comment]
@amaltaro, Alan, I ran the few tests it seems it is working as intended but code ends up a little complicate than I was hoping for. Please review.

File: src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py
Comment (amaltaro) [inline src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py]: typo

Comment (amaltaro) [reply src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py]: Maybe also add a docstring saying that this method decides whether a job can be submitted or not.

File: src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py
Comment (amaltaro) [inline src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py]: typo

File: src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py
Comment (amaltaro) [inline src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py]: more typos

**Review by amaltaro — APPROVED** [review]
Logic looks good, Seangchan. It took me a while to digest all the details (evil jobStats...), but it seems to be doing what it's supposed to do.
Several aesthetic fixes though.

File: src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py
Comment (ticoann) [reply src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py]: the reason I did max(totalPendingSlots, initTotalPending) to handle when 
totalPendingSlots >  initTotalPending. Lets say totalPendingSlots = 1000 and initTotalPending is 100 (which is current pending jobs in the system). We don't want to just submit just 20 percent in that case we want to submit up to the totalPendingSlots. 
I shouldn't add 20 percent extra in that case, but code will be more complicated and I think it is OK to overflow.

**Comment by ticoann** [comment]
Thanks Alan, I will fix all the typos.

File: src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py
Comment (amaltaro) [reply src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py]: From L543, isn't initTotalPending == totalPendingSlots ? Well, at least for the first time we iterate over a specific site. Actually, to me it seems it's always the same, since we don't do anything else with the pending slots (+1 or -1 etc) inside the submitter algo.

Comment (ticoann) [reply src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py]: Yes you're right. You found a bug. I was meant to put initiTotalPedning as initial totalPendingJobs not totalPendingSlots. Thanks I will update the patch.

**Comment by ticoann** [comment]
Alan could you review again. I fixed what you pointed.

File: src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py
Comment (amaltaro) [inline src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py]: I failed to see any difference :-) Maybe next time make a new commit only to make it clear what changed (and merge after you get the green light).

Anyways, isn't this line L546 == L536? And L550 == L542?

File: src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py
Comment (amaltaro) [inline src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py]: not sure if you forgot to replace it by logging.exception or just didn't want to. Ping'ing you again just to be sure.

**Review by amaltaro — COMMENTED** [review]
The max thing is still not clear to me though

File: src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py
Comment (ticoann) [reply src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py]: Of course I forgot. :-)

File: src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py
Comment (ticoann) [reply src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py]: Ah I need to change my glasses.  The difference was before it was total_pending_slot. Now it is total_pending_jobs

**Comment by ticoann** [comment]
About the max, I am still confused too. :-)
Here are the all the use cases we have handle.
1. We have to submit jobs regardless number of pending jobs with certain fraction of pending slots if the priority of jobs higher than pending jobs. 
2. We have to submit jobs up to pending slots if current pending jobs are fewer than pending slot regardless in fraction defined.

so without max(pending_job, pending_slot)
If we choose pending_slot (without max). it won't get to submit the jobs if pending jobs are more than 20 % of pending slots.

If we choose pending job (without max), it submits not enough jobs if the pending jobs are much smaller than the pending slots. (only 20% of pending slot) - Of course, next cycle it will put more jobs but we have to wait until next cycle.

**Comment by amaltaro** [comment]
test this please

**Comment by ticoann** [comment]
test this please

**Comment by ticoann** [comment]
@amaltaro, Alan I added fix for #8284 as well. Please review.

**Comment by ticoann** [comment]
@amaltaro, Alan commit is correctly updated now. Could you review?

File: src/python/WMComponent/JobSubmitter/JobSubmitAPI.py
Comment (amaltaro) [inline src/python/WMComponent/JobSubmitter/JobSubmitAPI.py]: typos :)

File: src/python/WMComponent/JobSubmitter/JobSubmitAPI.py
Comment (amaltaro) [inline src/python/WMComponent/JobSubmitter/JobSubmitAPI.py]: Isn't it WMCore.WMBS instead?

File: src/python/WMComponent/JobSubmitter/JobSubmitAPI.py
Comment (amaltaro) [inline src/python/WMComponent/JobSubmitter/JobSubmitAPI.py]: I wonder whether it really returns None in case it fails to fetch that parameter...
Anyways, my point is that this casting to int will crash if the returned value is None (my mistake btw :)).
How about we cast after the if below or in the L23?

File: src/python/WMComponent/JobSubmitter/JobSubmitAPI.py
Comment (amaltaro) [inline src/python/WMComponent/JobSubmitter/JobSubmitAPI.py]: Minor, but.... how about renaming it to something like 'availableScheddSlots'?

File: src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py
Comment (amaltaro) [inline src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py]: And now I see your point with this `max`. This is needed for cases where higher prio jobs come on top of other higher prio and so on. Still, considering 20% of the original pending slots threshold. You were right ;)
PS.: typos in the comments above.

File: src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py
Comment (amaltaro) [inline src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py]: just to minimize these assignments (that twists my brain), how about assigning this expression directly to `taskPendingThreshold`?
Same for the line above, assigning directly to `totalPendingThreshold` instead of `pendingTotalOverflow`.

File: src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py
Comment (amaltaro) [inline src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py]: Just to keep it consistent, how about renaming it from `NoTaskSlot` to `NoTaskRunningSlot`?

File: src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py
Comment (amaltaro) [inline src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py]: Since we already get the aux db config in L815. How about we move this 
`self.condorOverflowFraction = agentConfig.get("CondorOverflowFraction", 0.2)`
to L817?

File: src/python/WMComponent/WorkQueueManager/WorkQueueManagerWorkPoller.py
Comment (amaltaro) [inline src/python/WMComponent/WorkQueueManager/WorkQueueManagerWorkPoller.py]: dup msg

File: src/python/WMComponent/WorkQueueManager/WorkQueueManagerWorkPoller.py
Comment (amaltaro) [inline src/python/WMComponent/WorkQueueManager/WorkQueueManagerWorkPoller.py]: Just to avoid us checking all these constraints, how about we add one info (or warning) log per condition? Then we can remove this log line I think:
https://github.com/ticoann/WMCore/blob/a2f3245d72b595e040bcbe2760ef69c9ba55e903/src/python/WMComponent/WorkQueueManager/WorkQueueManagerWorkPoller.py#L94

**Review by amaltaro — APPROVED** [review]
Looks much better now, Seangchan!
I still have some suggestions though that will make the code cleaner and more readable, besides that, it's good to go (provided you did some simple tests :))

File: src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py
Comment (ticoann) [reply src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py]: Actually L815 need to be moved here 781 otherwise Tier0 agent will crash.

**Comment by ticoann** [comment]
Thank  you very much Alan for through review. I changed everything (I hope) as you suggested except one thing. I rearrange the aux db call so Tier0 Agent won't crash, also Tier0 is not checking the same condition.  Could you review again?

**Comment by ticoann** [comment]
test this please

**Comment by ticoann** [comment]
@amaltaro, Alan let me know 4th commit is OK. especially for Tier0 values

**Comment by ticoann** [comment]
@hufnagel, Dirk, FYI, this change will effect Tier0 Agent (in minimal way)

https://github.com/dmwm/WMCore/pull/8288/files#diff-63fe4938fe7d571728e9e8e9d3e0a1afR770

Basically it won't submit the job when schedd limit is reached. MAX_JOBS_PER_OWNER
Is this OK?

File: src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py
Comment (amaltaro) [inline src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py]: typo :)

File: src/python/WMComponent/WorkQueueManager/WorkQueueManagerWorkPoller.py
Comment (amaltaro) [inline src/python/WMComponent/WorkQueueManager/WorkQueueManagerWorkPoller.py]: Yes, I like it!

File: src/python/WMComponent/JobSubmitter/JobSubmitAPI.py
Comment (amaltaro) [inline src/python/WMComponent/JobSubmitter/JobSubmitAPI.py]: and since you'll have to fix the typo below, how about "executiing" and "differece"

**Review by amaltaro — CHANGES_REQUESTED** [review]
It looks like we are competing to see who makes more commits and get more comments on our PRs :)

Seangchan, code looks good, you just need to fix a typo in a dict key (and why not a couple more in the comments..) and we're good.

**Comment by hufnagel** [comment]
What is the default for MAX_JOBS_PER_OWNER ? 100k ? If so there shouldn't be an issue for the Tier0.

**Comment by amaltaro** [comment]
yes, 100k

**Comment by ticoann** [comment]
test this please

**Comment by ticoann** [comment]
test this please

**Pull request merged**
